README.md

```diff
@@ -50,4 +50,4 @@ reasoning, to this markdown file.
 
 ## Answer
 My code first starts with four constant operations, so we can just ignore those.
-Then we have a for loop that goes from $0$ to $n$ where $n$ is the number of cites in our matrix so $n$ so far. Nested in it, it has a for loop that fills an array from $0$ to $n$. We then have the call to our recursive tsp_hk, which has two for loops in it. The first for loop goes from $0$ to $n$. The second for loop goes from $0$ to $n-1$ because it is looping through all the new cities which contains all the old cities minus the start, but as it's a constant we can throw it away. So as we are iterating through all possible subsets of our original matrix of cities, we have $2^n$. Additionally, for each permutation we are doing $n^2$ work so, in total, we are doing $2^n n^2$ work. So our worst possible time complexity is $\Theta(2^n n^2)$.
+Then we have a for loop that goes from $0$ to $n$ where $n$ is the number of cites in our matrix so $n$ so far. Nested in it, it has a for loop that fills an array from $0$ to $n$. We then have the call to our recursive tsp_hk. The code first has a constant time operation then it has the first for loop goes which goes from $0$ to $n$. It then has another loop that goes from $0$ to $n-1$ because it is looping through all the new cities which contains all the old cities minus the start, but as it's a constant we can throw it away. It then some more constant time operations. So, as we are effectivly looping through every possible subset of cities the total runtime is $2^n$. So our worst possible time complexity is $\Theta(2^n)$.
```

Comment:
...but you're also using the start city as the key. Doesn't that affect the runtime?

Reply:
Yeah, so I would need to modify it to be something along the lines of $2^n$ multiplied by $log (n)$ as that's how fast our cache is filling up and in the case of a cache hit we just return immediately? Am I on the right track or still far off?

Reply:
Where does the $\log n$ come from?

Reply:
I was thinking that because every time you go through if it's not in the cache, you would add it and if it's in the cache it would subtract from the total possibilities. So if it's growing at $2^n$ you would have the inverse of possibilities left.

Reply:
Hmm, not following you here. The number of subsets left would be $2^n - $ however many you tried.

Reply:
So you would have something like $2^n - x$ where x is the number of subsets tried so far. How would you tie that back, or would you not need to?

Reply:
It doesn't really matter, because all you need to worry about is how many things would be in the cache in the worst case. And it's not just the subsets, but also the start cities.

Reply:
So we have $n$ cities and $2^n$ possible subsets. If we add memoization and it isn't in the cache we need to compute it which takes $n^2$. If its in the cache then we can just return immediately so its just a constant. So when put together would something like $\Theta(n * 2^n * (C + n^2))$ where $C$ is the constant factor of the cache hit and returning immediately. Is that closer or am I still far off?

Reply:
Not sure if I'm following you. How many different keys for the cache can there be?

Reply:
If there are $n$ cities and $2^n$ possible subsets then would it be 2^(n-1) possible cities in the cache?

Reply:
But the key you're using to index into the cache also uses the start city.

Reply:
So you would have an additional $n$ right? Im just not really sure where to put it.

Reply:
Comes back to the question of how many different keys you can have -- what effect does the start city have on the key?

Reply:
Using the start city as a part of the key stops the program from recalculating the shortest path with the same subset and same start city. So it makes sure that we don't do a bunch of redundant calculations.

Reply:
Ok great! But then you have more than n! possible keys, don't you?

Reply:
Yes, as we are trading size for speed.

Reply:
If we fill up every key, we would get a memory space complexity of $\Theta(n*2^n)$ because we create a cache entry for every combination of subset city and the start city. We then also have the time complexity which would be $\Theta(n * 2^n)$ because of the dynamic programming as opposed to the non dynamic complexity of $\Theta(n^2 * 2^n)$.

Reply:
That sounds like it's going in the right direction.